sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -75,23 +75,51 @@ ur_exp_command_buffer_handle_t_::ur_exp_command_buffer_handle_t_(
     ZeStruct<ze_command_list_desc_t> ZeDesc,
     const ur_exp_command_buffer_desc_t *Desc)
     : Context(hContext), Device(hDevice), ZeCommandList(CommandList),
-      ZeCommandListDesc(ZeDesc), QueueProperties() {
+      ZeCommandListDesc(ZeDesc), QueueProperties(), SyncPoints(),
+      NextSyncPoint(0), CommandListMap() {
   hContext->RefCount.increment();
   // TODO: Do we actually need the queue properties? Removed from the UR feature
   // for now.
 }
 
+// The ur_exp_command_buffer_handle_t_ destructor release all the memory objects
+// allocated for command_buffer managment
 ur_exp_command_buffer_handle_t_::~ur_exp_command_buffer_handle_t_() {
+  // Release the memory allocated to the Contexr stored in the command_buffer
+  if (Context->RefCount.decrementAndTest())
+    ZE_CALL_NOCHECK(zeContextDestroy, (Context->ZeContext));
+
+  // Release the memory allocated to the CommandList stored in the
+  // command_buffer
   if (ZeCommandList) {
     ZE_CALL_NOCHECK(zeCommandListDestroy, (ZeCommandList));
   }
+
+  // Release additional signal and wait events used by command_buffer
   if (SignalEvent) {
-    SignalEvent->RefCount.decrementAndTest();
+    CleanupCompletedEvent(SignalEvent, false);
+    urEventReleaseInternal(SignalEvent);
   }
   if (WaitEvent) {
-    WaitEvent->RefCount.decrementAndTest();
+    CleanupCompletedEvent(WaitEvent, false);
+    urEventReleaseInternal(WaitEvent);
+  }
+
+  // Release events added to the command_buffer
+  if (GetNextSyncPoint() > 0) {
```

Comment:
```suggestion
  if (SyncPoints.size()) {
```
I think this is a bit clearer, though functionally should be the same. Unless we were to overflow on the sync points I suppose!

Reply:
Do we need this size check at? If `SyncPoints` is empty we just won't do any iterations of the loop